WalletWasabi/Tor/Control/TorControlClient.cs

```diff
@@ -30,11 +35,50 @@ internal TorControlClient(PipeReader pipeReader, PipeWriter pipeWriter)
 			TcpClient = null;
 			PipeReader = pipeReader;
 			PipeWriter = pipeWriter;
+			ReaderCts = new();
+
+			SyncChannel = Channel.CreateUnbounded<TorControlReply>(Options);
+			AsyncChannels = new List<Channel<TorControlReply>>();
+			ReaderLoopTask = Task.Run(ReaderLoopAsync);
+			MessageLock = new();
 		}
 
 		private TcpClient? TcpClient { get; }
 		private PipeReader PipeReader { get; }
 		private PipeWriter PipeWriter { get; }
+		private CancellationTokenSource ReaderCts { get; }
+		private Task ReaderLoopTask { get; }
+
+		/// <summary>Channel only for synchronous replies from Tor control.</summary>
+		/// <remarks>Typically, there is at most one message in the channel at a time.</remarks>
+		private Channel<TorControlReply> SyncChannel { get; }
+
+		/// <summary>Guards <see cref="AsyncChannels"/>.</summary>
+		private object AsyncChannelsLock { get; } = new();
+
+		/// <summary>Channel only for <see cref="StatusCode.AsynchronousEventNotify"/> events.</summary>
+		/// <remarks>
+		/// Guarded by <see cref="AsyncChannelsLock"/>.
+		/// <para>This list should be used only in a copy-on-write way to avoid iterating a modified list.</para>
+		/// </remarks>
+		private List<Channel<TorControlReply>> AsyncChannels { get; set; }
+
+		/// <summary>Lock to when sending a request to Tor control and waiting for a reply.</summary>
+		/// <remarks>Tor control protocol does not provide a foolproof way to recognize that a response belongs to a request.</remarks>
+		private AsyncLock MessageLock { get; }
+
+		/// <summary>Number of subscribers that currently listen to Tor's async events using <see cref="ReadEventsAsync"/>.</summary>
+		/// <remarks>Mainly for tests.</remarks>
+		internal int SubscriberCount
+		{
+			get
+			{
+				lock (AsyncChannelsLock)
+				{
+					return AsyncChannels.Count;
```

Comment:
`SyncChannel` is not part of the ` AsyncChannels`, so the `SubscriberCount` shouldn't return `+ 1`?

Reply:
Conceptually the current impl is better.

Reply:
Yes, that's correct.